<h2>Results summary for Hisat Stringtie and Ballgown workflow</h2>
If you download the whole gzipped file (e.g. jobid.tar.gz), please unpack it with command such as
"tar xvzf jobid.tar.gz", this will extract all the computed results in your current
directory.

The following are the files and directories you may see:

NGS-config
    Configuration file and parameters

NGS-samples
    List of samples

NGS-env
    Environmental variables from the server where the jobs are executed

Sample_name
    This is a directory where results for this sample are stored.
    This directory can have several sub-directories with results from
    individual tools (e.g. hisat, stringtie, ballgown).
    Within each sub-directory, all necessary results
    can be found. 
    In the folder of qc-hisat-stringtie:
        sorted.bam: mapping results
        transcripts.gtf: transcripts assembling results
    In the folder of ballgown:
        ballgown.gtf:
        e_data.ctab: exon-level expression measurements
        i_data.ctab: intron-level expression measurements
        t_data.ctab: transcript-level expression measurements
        i2t.ctab: table with two columns, i_id and t_id
        e2t.ctab: table with two columns, e_id and t_id

Merged_transcripts
    Merged transcripts by stringtie from transcripts from all samples
        stringtie-merge/stringtie_merged.gtf

Group1_vs_Group2 (or Group1_vs_Group3 etc)
    ballgown results between samples of group1 and 2
     *  genes_results.txt: differential analysis on genes level 
     *  transcripts_results.txt: differential analysis on transcripts level
        two_group_sample.txt 
        rnaseq_ballgown.R : R script for ballgown analysis
        rnaseq_ballgown.Rout : R output for ballgown analysis
        ballgown 

WF-sh
    This is a directory that stores the actual shell scripts, all the parameters
    can be found here.

* indicates the files that you may be most interested in.
