<h2>Results summary for STAR alignment and post analysis workflow</h2>
If you download the whole gzipped file (e.g. jobid.tar.gz), please unpack it with command such as
"tar xvzf jobid.tar.gz", this will extract all the computed results in your current
directory, following are the files and directories you may see:

NGS-config
    Configuration file and parameters

NGS-samples
    List of samples 

NGS-env
    Environmental variables from the server where the jobs were executed

Sample_Name
    This is a directory where results for this sample are stored.
    This directory can have several sub-directories with results from
    individual tools. Within this directory, all necessary results can be found, e.g.
        starAligned.sortedByCoord.out.bam 
        starAligned.toTranscriptome.out.bam 
        starLog.final.out 
        starLog.out 
        starLog.progress.out 
        starReadsPerGene.out.tab 
        starSJ.out.tab 
        starSignal.Unique.str1.out.bedgraph 
        starSignal.Unique.str2.out.bedgraph 
        starSignal.UniqueMultiple.str1.out.bedgraph 
        starSignal.UniqueMultiple.str2.out.bedgraph 
        RSEM.genes.results 
        RSEM.isoforms.results 

Sample-all-transcript-matrix-group
    Post analysis of differentially expressed transcripts
    *   post-RSEM-group/genes.counts.matrix
        post-RSEM-group/genes.not_cross_norm.fpkm.tmp
        post-RSEM-group/genes.not_cross_norm.fpkm.tmp.TMM_info.txt
        post-RSEM-group/genes.TMM.fpkm.matrix
    *   post-RSEM-group/trans.counts.matrix
        post-RSEM-group/trans.not_cross_norm.fpkm.tmp
        post-RSEM-group/trans.not_cross_norm.fpkm.tmp.TMM_info.txt
        post-RSEM-group/trans.TMM.fpkm.matrix
        post-RSEM-group/WF.complete.date
        post-RSEM-group/WF.cpu
        post-RSEM-group/WF.start.date
        post-RSEM-group/genes-diff/clusters_fixed_P_50.heatmap.heatmap.pdf
    *   post-RSEM-group/genes-diff/genes.counts.matrix.Group1_vs_Group2.edgeR.DE_results
        post-RSEM-group/genes-diff/genes.counts.matrix.Group1_vs_Group2.edgeR.DE_results.MA_n_Volcano.pdf
        post-RSEM-group/genes-diff/genes.counts.matrix.Group1_vs_Group2.edgeR.DE_results.P1e-3_C2.Group1-UP.subset
        post-RSEM-group/genes-diff/genes.counts.matrix.Group1_vs_Group2.edgeR.DE_results.P1e-3_C2.Group2-UP.subset
        post-RSEM-group/genes-diff/genes.counts.matrix.Group1_vs_Group2.edgeR.DE_results.samples
        post-RSEM-group/genes-diff/genes.counts.matrix.Group1_vs_Group2.Group1.vs.Group2.EdgeR.Rscript
    *   post-RSEM-group/genes-diff/genes-diff.matrix
        post-RSEM-group/genes-diff/genes-diff.matrix.log2.centered.dat
        post-RSEM-group/genes-diff/genes-diff.matrix.log2.centered.genes_vs_samples_heatmap.pdf
        post-RSEM-group/genes-diff/genes-diff.matrix.log2.dat
        post-RSEM-group/genes-diff/genes-diff.matrix.log2.sample_cor.dat
        post-RSEM-group/genes-diff/genes-diff.matrix.log2.sample_cor_matrix.pdf
        post-RSEM-group/genes-diff/genes-diff.matrix.R
        post-RSEM-group/genes-diff/genes-diff.matrix.RData
        post-RSEM-group/genes-diff/numDE_feature_counts.P1e-3_C2.matrix
        post-RSEM-group/genes-diff/__tmp_define_clusters.R
        post-RSEM-group/trans-diff/clusters_fixed_P_50.heatmap.heatmap.pdf
        post-RSEM-group/trans-diff/numDE_feature_counts.P1e-3_C2.matrix
        post-RSEM-group/trans-diff/__tmp_define_clusters.R
    *   post-RSEM-group/trans-diff/trans.counts.matrix.Group1_vs_Group2.edgeR.DE_results
        post-RSEM-group/trans-diff/trans.counts.matrix.Group1_vs_Group2.edgeR.DE_results.MA_n_Volcano.pdf
        post-RSEM-group/trans-diff/trans.counts.matrix.Group1_vs_Group2.edgeR.DE_results.P1e-3_C2.Group1-UP.subset
        post-RSEM-group/trans-diff/trans.counts.matrix.Group1_vs_Group2.edgeR.DE_results.P1e-3_C2.Group2-UP.subset
        post-RSEM-group/trans-diff/trans.counts.matrix.Group1_vs_Group2.edgeR.DE_results.samples
        post-RSEM-group/trans-diff/trans.counts.matrix.Group1_vs_Group2.Group1.vs.Group2.EdgeR.Rscript
    *   post-RSEM-group/trans-diff/trans-diff.matrix
        post-RSEM-group/trans-diff/trans-diff.matrix.log2.centered.dat
        post-RSEM-group/trans-diff/trans-diff.matrix.log2.centered.genes_vs_samples_heatmap.pdf
        post-RSEM-group/trans-diff/trans-diff.matrix.log2.dat
        post-RSEM-group/trans-diff/trans-diff.matrix.log2.sample_cor.dat
        post-RSEM-group/trans-diff/trans-diff.matrix.log2.sample_cor_matrix.pdf
        post-RSEM-group/trans-diff/trans-diff.matrix.R
        post-RSEM-group/trans-diff/trans-diff.matrix.RData
WF-sh
    This is a directory that stores the actual shell scripts, all the parameters
    can be found here.

* indicated the files that you may be most interested in.
